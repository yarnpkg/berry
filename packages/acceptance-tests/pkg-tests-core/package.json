{
  "name": "pkg-tests-core",
  "private": true,
  "license": "BSD-2-Clause",
  "main": "./sources/index.ts",
  "devDependencies": {
    "@types/finalhandler": "1.1.0",
    "@types/invariant": "^2.2.30",
    "@types/klaw": "^2.0.0",
    "@types/minimatch": "^3.0.3",
    "@types/pem": "^1",
    "@types/semver": "^7.1.0",
    "@types/serve-static": "1.13.3",
    "@types/tar-fs": "^1.16.1",
    "@yarnpkg/cli": "workspace:*",
    "@yarnpkg/core": "workspace:*",
    "@yarnpkg/fslib": "workspace:*",
    "@yarnpkg/parsers": "workspace:*",
    "finalhandler": "^1.1.2",
    "invariant": "^2.2.4",
    "klaw": "^2.1.1",
    "minimatch": "^3.0.4",
    "pem": "^1.14.4",
    "pkg-tests-fixtures": "workspace:*",
    "semver": "^7.1.2",
    "serve-static": "^1.14.1",
    "super-resolve": "^1.0.0",
    "tar-fs": "^1.16.0"
  },
  "repository": {
    "type": "git",
    "url": "ssh://git@github.com/yarnpkg/berry.git",
    "directory": "packages/acceptance-tests/pkg-tests-core"
  },
  "engines": {
    "node": ">=12 <14 || 14.2 - 14.9 || >14.10.0"
  }
}
