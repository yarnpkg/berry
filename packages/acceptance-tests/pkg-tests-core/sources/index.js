const exec = require('./utils/exec');
const fs = require('./utils/fs');
const misc = require('./utils/misc');
const tests = require('./utils/tests');

module.exports = {exec, fs, misc, tests};
