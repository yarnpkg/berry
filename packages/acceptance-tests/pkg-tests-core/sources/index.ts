import './utils/makeTemporaryEnv';
import './utils/makeTemporaryMonorepoEnv';
import * as exec  from './utils/exec';
import * as fs    from './utils/fs';
import * as misc  from './utils/misc';
import * as tests from './utils/tests';
import * as yarn  from './utils/yarn';

export {exec, fs, misc, tests, yarn};
