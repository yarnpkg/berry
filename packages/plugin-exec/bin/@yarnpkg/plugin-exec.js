/* eslint-disable */
module.exports = {
name: "@yarnpkg/plugin-exec",
factory: function (require) {
var plugin;plugin=(()=>{"use strict";var e={230:(e,t,r)=>{r.r(t),r.d(t,{execUtils:()=>a,default:()=>p});var a={};r.r(a),r.d(a,{loadGeneratorFile:()=>l,makeLocator:()=>c,makeSpec:()=>i,parseSpec:()=>s});var o=r(966),n=r(688);function s(e){const{params:t,selector:r}=o.structUtils.parseRange(e),a=n.npath.toPortablePath(r);return{parentLocator:t&&"string"==typeof t.locator?o.structUtils.parseLocator(t.locator):null,path:a}}function i({parentLocator:e,path:t,generatorHash:r,protocol:a}){const n=null!==e?{locator:o.structUtils.stringifyLocator(e)}:{},s=void 0!==r?{hash:r}:{};return o.structUtils.makeRange({protocol:a,source:t,selector:t,params:{...s,...n}})}function c(e,{parentLocator:t,path:r,generatorHash:a,protocol:n}){return o.structUtils.makeLocator(e,i({parentLocator:t,path:r,generatorHash:a,protocol:n}))}async function l(e,t,r){const{parentLocator:a,path:s}=o.structUtils.parseFileStyleRange(e,{protocol:t}),i=n.ppath.isAbsolute(s)?{packageFs:new n.CwdFS(n.PortablePath.root),prefixPath:n.PortablePath.dot,localPath:n.PortablePath.root}:await r.fetcher.fetch(a,r),c=i.localPath?{packageFs:new n.CwdFS(n.PortablePath.root),prefixPath:n.ppath.relative(n.PortablePath.root,i.localPath)}:i;i!==c&&i.releaseFs&&i.releaseFs();const l=c.packageFs,p=n.ppath.join(c.prefixPath,s);return await l.readFilePromise(p,"utf8")}const p={fetchers:[class{supports(e,t){return!!e.reference.startsWith("exec:")}getLocalPath(e,t){const{parentLocator:r,path:a}=o.structUtils.parseFileStyleRange(e.reference,{protocol:"exec:"});if(n.ppath.isAbsolute(a))return a;const s=t.fetcher.getLocalPath(r,t);return null===s?null:n.ppath.resolve(s,a)}async fetch(e,t){const r=t.checksums.get(e.locatorHash)||null,[a,n,s]=await t.cache.fetchPackageFromCache(e,r,{onHit:()=>t.report.reportCacheHit(e),onMiss:()=>t.report.reportCacheMiss(e),loader:()=>this.fetchFromDisk(e,t),skipIntegrityCheck:t.skipIntegrityCheck});return{packageFs:a,releaseFs:n,prefixPath:o.structUtils.getIdentVendorPath(e),localPath:this.getLocalPath(e,t),checksum:s}}async fetchFromDisk(e,t){const r=await l(e.reference,"exec:",t);return n.xfs.mktempPromise(async a=>{const s=n.ppath.join(a,"generator.js");return await n.xfs.writeFilePromise(s,r),n.xfs.mktempPromise(async r=>{if(await this.generatePackage(r,e,s,t),!n.xfs.existsSync(n.ppath.join(r,"build")))throw new Error("The script should have generated a build directory");return await o.tgzUtils.makeArchiveFromDirectory(n.ppath.join(r,"build"),{prefixPath:o.structUtils.getIdentVendorPath(e),compressionLevel:t.project.configuration.get("compressionLevel")})})})}async generatePackage(e,t,r,a){return await n.xfs.mktempPromise(async s=>{const i=await o.scriptUtils.makeScriptEnv({project:a.project,binFolder:s}),c=n.ppath.join(e,"runtime.js");return await n.xfs.mktempPromise(async a=>{const s=n.ppath.join(a,"buildfile.log"),l=n.xfs.createWriteStream(s),p=l,u=n.ppath.join(e,"generator"),h=n.ppath.join(e,"build");await n.xfs.mkdirPromise(u),await n.xfs.mkdirPromise(h);const f={tempDir:n.npath.fromPortablePath(u),buildDir:n.npath.fromPortablePath(h),locator:o.structUtils.stringifyLocator(t)};await n.xfs.writeFilePromise(c,`\n          // Expose 'Module' as a global variable\n          Object.defineProperty(global, 'Module', {\n            get: () => require('module'),\n            configurable: true,\n            enumerable: false,\n          });\n\n          // Expose non-hidden built-in modules as global variables\n          for (const name of Module.builtinModules.filter((name) => name !== 'module' && !name.startsWith('_'))) {\n            Object.defineProperty(global, name, {\n              get: () => require(name),\n              configurable: true,\n              enumerable: false,\n            });\n          }\n\n          // Expose the 'execEnv' global variable\n          Object.defineProperty(global, 'execEnv', {\n            value: {\n              ...${JSON.stringify(f)},\n            },\n            enumerable: true,\n          });\n        `);let d=i.NODE_OPTIONS||"";d=d.replace(/\s*--require\s+\S*\.pnp\.c?js\s*/g," ").trim(),i.NODE_OPTIONS=d,l.write(`# This file contains the result of Yarn generating a package (${o.structUtils.stringifyLocator(t)})\n`),l.write("\n");const{code:g}=await o.execUtils.pipevp(process.execPath,["--require",n.npath.fromPortablePath(c),n.npath.fromPortablePath(r),o.structUtils.stringifyIdent(t)],{cwd:e,env:i,stdin:null,stdout:l,stderr:p});if(0!==g)throw n.xfs.detachTemp(a),new Error(`Package generation failed (exit code ${g}, logs can be found here: ${s})`)})})}}],resolvers:[class{supportsDescriptor(e,t){return!!e.range.startsWith("exec:")}supportsLocator(e,t){return!!e.reference.startsWith("exec:")}shouldPersistResolution(e,t){return!1}bindDescriptor(e,t,r){return o.structUtils.bindDescriptor(e,{locator:o.structUtils.stringifyLocator(t)})}getResolutionDependencies(e,t){return[]}async getCandidates(e,t,r){if(!r.fetchOptions)throw new Error("Assertion failed: This resolver cannot be used unless a fetcher is configured");const{path:a,parentLocator:n}=s(e.range);if(null===n)throw new Error("Assertion failed: The descriptor should have been bound");const i=await l(o.structUtils.makeRange({protocol:"exec:",source:a,selector:a,params:{locator:o.structUtils.stringifyLocator(n)}}),"exec:",r.fetchOptions);return[c(e,{parentLocator:n,path:a,generatorHash:o.hashUtils.makeHash("1",i).slice(0,6),protocol:"exec:"})]}async getSatisfying(e,t,r){return null}async resolve(e,t){if(!t.fetchOptions)throw new Error("Assertion failed: This resolver cannot be used unless a fetcher is configured");const r=await t.fetchOptions.fetcher.fetch(e,t.fetchOptions),a=await o.miscUtils.releaseAfterUseAsync(async()=>await o.Manifest.find(r.prefixPath,{baseFs:r.packageFs}),r.releaseFs);return{...e,version:a.version||"0.0.0",languageName:t.project.configuration.get("defaultLanguageName"),linkType:o.LinkType.HARD,dependencies:a.dependencies,peerDependencies:a.peerDependencies,dependenciesMeta:a.dependenciesMeta,peerDependenciesMeta:a.peerDependenciesMeta,bin:a.bin}}}]}},966:e=>{e.exports=require("@yarnpkg/core")},688:e=>{e.exports=require("@yarnpkg/fslib")}},t={};function r(a){if(t[a])return t[a].exports;var o=t[a]={exports:{}};return e[a](o,o.exports,r),o.exports}return r.d=(e,t)=>{for(var a in t)r.o(t,a)&&!r.o(e,a)&&Object.defineProperty(e,a,{enumerable:!0,get:t[a]})},r.o=(e,t)=>Object.prototype.hasOwnProperty.call(e,t),r.r=e=>{"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},r(230)})();
return plugin;
}
};