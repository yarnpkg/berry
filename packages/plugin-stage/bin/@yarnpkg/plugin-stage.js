module.exports={name:"@yarnpkg/plugin-stage",factory:function(t){return function(t){var e={};function n(o){if(e[o])return e[o].exports;var a=e[o]={i:o,l:!1,exports:{}};return t[o].call(a.exports,a,a.exports,n),a.l=!0,a.exports}return n.m=t,n.c=e,n.d=function(t,e,o){n.o(t,e)||Object.defineProperty(t,e,{enumerable:!0,get:o})},n.r=function(t){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(t,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(t,"__esModule",{value:!0})},n.t=function(t,e){if(1&e&&(t=n(t)),8&e)return t;if(4&e&&"object"==typeof t&&t&&t.__esModule)return t;var o=Object.create(null);if(n.r(o),Object.defineProperty(o,"default",{enumerable:!0,value:t}),2&e&&"string"!=typeof t)for(var a in t)n.d(o,a,function(e){return t[e]}.bind(null,a));return o},n.n=function(t){var e=t&&t.__esModule?function(){return t.default}:function(){return t};return n.d(e,"a",e),e},n.o=function(t,e){return Object.prototype.hasOwnProperty.call(t,e)},n.p="",n(n.s=4)}([function(e,n){e.exports=t("@yarnpkg/fslib")},function(e,n){e.exports=t("@yarnpkg/core")},function(e,n){e.exports=t("clipanion")},function(e,n){e.exports=t("@yarnpkg/cli")},function(t,e,n){"use strict";n.r(e);function o(t,e,n,o){var a,i=arguments.length,r=i<3?e:null===o?o=Object.getOwnPropertyDescriptor(e,n):o;if("object"==typeof Reflect&&"function"==typeof Reflect.decorate)r=Reflect.decorate(t,e,n,o);else for(var s=t.length-1;s>=0;s--)(a=t[s])&&(r=(i<3?a(r):i>3?a(e,n,r):a(e,n))||r);return i>3&&r&&Object.defineProperty(e,n,r),r}var a,i=n(3),r=n(1),s=n(0),c=n(2);async function l(t,{marker:e}){do{if(s.xfs.existsSync(s.ppath.join(t,e)))return t;t=s.ppath.dirname(t)}while("/"!==t);return null}function u(t,e){let n=0,o=0;for(const a of t)"wip"!==a&&(e.test(a)?n+=1:o+=1);return n>=o}!function(t){t[t.CREATE=0]="CREATE",t[t.DELETE=1]="DELETE",t[t.ADD=2]="ADD",t[t.REMOVE=3]="REMOVE",t[t.MODIFY=4]="MODIFY"}(a||(a={}));const p=new Map([[a.CREATE,"create"],[a.DELETE,"delete"],[a.ADD,"add"],[a.REMOVE,"remove"],[a.MODIFY,"update"]]);const f=11;async function d(t){const{code:e,stdout:n}=await r.execUtils.execvp("git",["log","-1","--pretty=format:%H"],{cwd:t});return 0===e?n.trim():null}async function m(t,e){const n=[],o=e.filter(t=>"package.json"===s.ppath.basename(t.path));for(const{action:e,path:i}of o){const o=s.ppath.relative(t,i);if(e===a.MODIFY){const e=await d(t),{stdout:s}=await r.execUtils.execvp("git",["show",`${e}:${o}`],{cwd:t,strict:!0}),c=await r.Manifest.fromText(s),l=await r.Manifest.fromFile(i),u=new Map([...l.dependencies,...l.devDependencies]),p=new Map([...c.dependencies,...c.devDependencies]);for(const[t,e]of p){const o=r.structUtils.stringifyIdent(e),i=u.get(t);i?i.range!==e.range&&n.push([a.MODIFY,`${o} to ${i.range}`]):n.push([a.REMOVE,o])}for(const[t,e]of u)p.has(t)||n.push([a.ADD,r.structUtils.stringifyIdent(e)])}else if(e===a.CREATE){const t=await r.Manifest.fromFile(i);t.name?n.push([a.CREATE,r.structUtils.stringifyIdent(t.name)]):n.push([a.CREATE,"a package"])}else{if(e!==a.DELETE)throw new Error("Assertion failed: Unsupported action type");{const e=await d(t),{stdout:i}=await r.execUtils.execvp("git",["show",`${e}:${o}`],{cwd:t,strict:!0}),s=await r.Manifest.fromText(i);s.name?n.push([a.DELETE,r.structUtils.stringifyIdent(s.name)]):n.push([a.DELETE,"a package"])}}}const{code:i,stdout:c}=await r.execUtils.execvp("git",["log",`-${f}`,"--pretty=format:%s"],{cwd:t});return function(t,e){const n=function(t){return t.useComponent?"chore(yarn): ":""}(t),o=[],a=e.slice().sort((t,e)=>t[0]-e[0]);for(;a.length>0;){const[e,n]=a.shift();let i=p.get(e);t.useUpperCase&&0===o.length&&(i=`${i[0].toUpperCase()}${i.slice(1)}`),t.useThirdPerson&&(i+="s");const r=[n];for(;a.length>0&&a[0][0]===e;){const[,t]=a.shift();r.push(t)}r.sort();let s=r.shift();1===r.length?s+=" (and one other)":r.length>1&&(s+=` (and ${r.length} others)`),o.push(`${i} ${s}`)}return`${n}${o.join(", ")}`}(function(t){return{useThirdPerson:u(t,/^(\w\(\w+\):\s*)?\w+s/),useUpperCase:u(t,/^(\w\(\w+\):\s*)?[A-Z]/),useComponent:u(t,/^\w\(\w+\):/)}}(0===i?c.split(/\n/g).filter(t=>""!==t):[]),n)}const h=[{findRoot:async t=>await l(t,{marker:Object(s.toFilename)(".git")}),async filterChanges(t,e,n){const{stdout:o}=await r.execUtils.execvp("git",["status","-s"],{cwd:t,strict:!0}),i=o.toString().split(/\n/g);return[].concat(...i.map(e=>{if(""===e)return[];const n=e.slice(0,3),o=s.ppath.resolve(t,e.slice(3));return"?? "===n&&e.endsWith("/")?function(t){const e=[],n=[t];for(;n.length>0;){const t=n.pop(),o=s.xfs.readdirSync(t);for(const a of o){const o=s.ppath.resolve(t,a);s.xfs.lstatSync(o).isDirectory()?n.push(o):e.push(o)}}return e}(o).map(t=>({action:a.CREATE,path:t})):" A "===n||"?? "===n?[{action:a.CREATE,path:o}]:" M "===n?[{action:a.MODIFY,path:o}]:" D "===n?[{action:a.DELETE,path:o}]:[]})).filter(t=>(function(t,{roots:e,names:n}){if(n.has(s.ppath.basename(t)))return!0;do{if(e.has(t))return!0;t=s.ppath.dirname(t)}while("/"!==t);return!1})(t.path,{roots:e,names:n}))},genCommitMessage:async(t,e)=>await m(t,e),async makeCommit(t,e,n){const o=e.map(t=>s.npath.fromPortablePath(t.path));await r.execUtils.execvp("git",["add","-N","--",...o],{cwd:t,strict:!0}),await r.execUtils.execvp("git",["commit","-m",`${n}\n\nCommit generated via \`yarn stage\`\n`,"--",...o],{cwd:t,strict:!0})},async makeReset(t,e){const n=e.map(t=>s.npath.fromPortablePath(t.path));await r.execUtils.execvp("git",["reset","HEAD","--",...n],{cwd:t,strict:!0})}},{findRoot:async t=>await l(t,{marker:Object(s.toFilename)(".hg")}),filterChanges:async(t,e,n)=>[],genCommitMessage:async(t,e)=>"",async makeCommit(t,e,n){},async makeReset(t,e){},async makeUpdate(t,e){}}];class g extends i.BaseCommand{constructor(){super(...arguments),this.commit=!1,this.reset=!1,this.update=!1,this.dryRun=!1}async execute(){const t=await r.Configuration.find(this.context.cwd,this.context.plugins),{project:e}=await r.Project.find(t,this.context.cwd),{driver:n,root:o}=await async function(t){let e=null,n=null;for(const o of h)if(null!==(n=await o.findRoot(t))){e=o;break}if(null===e||null===n)throw new c.UsageError("No stage driver has been found for your current project");return{driver:e,root:n}}(e.cwd),a=[t.get("bstatePath"),t.get("cacheFolder"),t.get("globalFolder"),t.get("virtualFolder"),t.get("yarnPath")];await t.triggerHook(t=>t.populateYarnPaths,e,t=>{a.push(t)});const i=new Set;for(const t of a)for(const e of y(o,t))i.add(e);const l=new Set([t.get("rcFilename"),t.get("lockfileFilename"),"package.json"]),u=await n.filterChanges(o,i,l),p=await n.genCommitMessage(o,u);if(this.dryRun)if(this.commit)this.context.stdout.write(`${p}\n`);else for(const t of u)this.context.stdout.write(`${s.npath.fromPortablePath(t.path)}\n`);else 0===u.length?this.context.stdout.write("No changes found!"):this.commit?await n.makeCommit(o,u,p):this.reset&&await n.makeReset(o,u)}}function y(t,e){const n=[];if(null===e)return n;for(;;){let o;(e===t||e.startsWith(`${t}/`))&&n.push(e);try{o=s.xfs.statSync(e)}catch(t){break}if(!o.isSymbolicLink())break;e=s.ppath.resolve(s.ppath.dirname(e),s.xfs.readlinkSync(e))}return n}g.usage=c.Command.Usage({description:"add all yarn files to your vcs",details:"\n      This command will add to your staging area the files belonging to Yarn (typically any modified `package.json` and `.yarnrc.yml` files, but also linker-generated files, cache data, etc). It will take your ignore list into account, so the cache files won't be added if the cache is ignored in a `.gitignore` file (assuming you use Git).\n\n      Running `--reset` will instead remove them from the staging area (the changes will still be there, but won't be committed until you stage them back).\n\n      Since the staging area is a non-existent concept in Mercurial, Yarn will always create a new commit when running this command on Mercurial repositories. You can get this behavior when using Git by using the `--commit` flag which will directly create a commit.\n    ",examples:[["Adds all modified project files to the staging area","yarn stage"],["Creates a new commit containing all modified project files","yarn stage --commit"]]}),o([c.Command.Boolean("-c,--commit")],g.prototype,"commit",void 0),o([c.Command.Boolean("-r,--reset")],g.prototype,"reset",void 0),o([c.Command.Boolean("-u,--update")],g.prototype,"update",void 0),o([c.Command.Boolean("-n,--dry-run")],g.prototype,"dryRun",void 0),o([c.Command.Path("stage")],g.prototype,"execute",null);const w={commands:[g]};e.default=w}])}};