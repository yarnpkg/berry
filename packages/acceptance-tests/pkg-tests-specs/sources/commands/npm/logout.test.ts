import {Filename, xfs} from '@yarnpkg/fslib';
import {yarn}          from 'pkg-tests-core';

const SPEC_RC_FILENAME = `.spec-yarnrc` as Filename;

const FAKE_FIRST_SCOPE = `first`;
const FAKE_SECOND_SCOPE = `second`;
const FAKE_THIRD_SCOPE = `third`;

const FAKE_REGISTRY_URL = `http://yarn.test.registry`;
const FAKE_PUBLISH_REGISTRY_URL = `https://npm.pkg.github.com`;
const FAKE_THIRD_REGISTRY_URL = `https://third.yarn.test.registry`;

const CLASSIC_SCOPE_SETTINGS = {
  npmAlwaysAuth: `true`,
};

const FAKE_REGISTRY_CREDENTIALS = {
  npmAuthIdent: `username:password`,
  npmAuthToken: `ffffffff-ffff-ffff-ffff-ffffffffffff`,
};

describe(`Commands`, () => {
  describe(`npm logout`, () => {
    it(
      `should logout a user from the default registry when no arguments are passed`,
      makeTemporaryEnv({}, {
        npmRegistryServer: FAKE_REGISTRY_URL,
        npmPublishRegistry: FAKE_PUBLISH_REGISTRY_URL,
      }, async ({path, run, source}) => {
        const homePath = await xfs.mktempPromise();

        await yarn.writeConfiguration(homePath, {
          npmRegistries: {
            [FAKE_REGISTRY_URL]: FAKE_REGISTRY_CREDENTIALS,
            [FAKE_PUBLISH_REGISTRY_URL]: FAKE_REGISTRY_CREDENTIALS,
            [FAKE_THIRD_REGISTRY_URL]: FAKE_REGISTRY_CREDENTIALS,
          },
          npmScopes: {
            [FAKE_FIRST_SCOPE]: FAKE_REGISTRY_CREDENTIALS,
          },
        }, {
          filename: SPEC_RC_FILENAME,
        });

        let code: number;
        let stdout: string;
        let stderr: string;

        try {
          // @ts-expect-error
          ({code, stdout, stderr} = await run(`npm`, `logout`, {
            env: {
              HOME: homePath,
              USERPROFILE: homePath,
              YARN_RC_FILENAME: SPEC_RC_FILENAME,
            },
          }));
        } catch (error) {
          ({code, stdout, stderr} = error);
        }

        await expect(yarn.readConfiguration(homePath, {
          filename: SPEC_RC_FILENAME,
        })).resolves.toStrictEqual({
          npmRegistries: {
            [FAKE_PUBLISH_REGISTRY_URL]: FAKE_REGISTRY_CREDENTIALS,
            [FAKE_THIRD_REGISTRY_URL]: FAKE_REGISTRY_CREDENTIALS,
          },
          npmScopes: {
            [FAKE_FIRST_SCOPE]: FAKE_REGISTRY_CREDENTIALS,
          },
        });

        expect({code, stdout, stderr}).toMatchSnapshot();
      })
    );

    it(
      `should logout a user from all registries and all scopes when the -A,--all flag is used`,
      makeTemporaryEnv({}, {
        npmRegistryServer: FAKE_REGISTRY_URL,
        npmPublishRegistry: FAKE_PUBLISH_REGISTRY_URL,
      }, async ({path, run, source}) => {
        const homePath = await xfs.mktempPromise();

        await yarn.writeConfiguration(homePath, {
          npmRegistries: {
            [FAKE_REGISTRY_URL]: FAKE_REGISTRY_CREDENTIALS,
            [FAKE_PUBLISH_REGISTRY_URL]: FAKE_REGISTRY_CREDENTIALS,
            [FAKE_THIRD_REGISTRY_URL]: FAKE_REGISTRY_CREDENTIALS,
          },
          npmScopes: {
            [FAKE_FIRST_SCOPE]: FAKE_REGISTRY_CREDENTIALS,
            [FAKE_SECOND_SCOPE]: CLASSIC_SCOPE_SETTINGS,
            [FAKE_THIRD_SCOPE]: {
              ...CLASSIC_SCOPE_SETTINGS,
              ...FAKE_REGISTRY_CREDENTIALS,
            },
          },
        }, {
          filename: SPEC_RC_FILENAME,
        });

        let code: number;
        let stdout: string;
        let stderr: string;

        try {
          // @ts-expect-error
          ({code, stdout, stderr} = await run(`npm`, `logout`, `--all`, {
            env: {
              HOME: homePath,
              USERPROFILE: homePath,
              YARN_RC_FILENAME: SPEC_RC_FILENAME,
            },
          }));
        } catch (error) {
          ({code, stdout, stderr} = error);
        }

        await expect(yarn.readConfiguration(homePath, {
          filename: SPEC_RC_FILENAME,
        })).resolves.toStrictEqual({
          npmScopes: {
            [FAKE_SECOND_SCOPE]: CLASSIC_SCOPE_SETTINGS,
            [FAKE_THIRD_SCOPE]: CLASSIC_SCOPE_SETTINGS,
          },
        });

        expect({code, stdout, stderr}).toMatchSnapshot();
      })
    );

    it(
      `should logout a user from the publish registry when the --publish flag is used`,
      makeTemporaryEnv({}, {
        npmRegistryServer: FAKE_REGISTRY_URL,
        npmPublishRegistry: FAKE_PUBLISH_REGISTRY_URL,
      }, async ({path, run, source}) => {
        const homePath = await xfs.mktempPromise();

        await yarn.writeConfiguration(homePath, {
          npmRegistries: {
            [FAKE_REGISTRY_URL]: FAKE_REGISTRY_CREDENTIALS,
            [FAKE_PUBLISH_REGISTRY_URL]: FAKE_REGISTRY_CREDENTIALS,
            [FAKE_THIRD_REGISTRY_URL]: FAKE_REGISTRY_CREDENTIALS,
          },
        }, {
          filename: SPEC_RC_FILENAME,
        });

        let code: number;
        let stdout: string;
        let stderr: string;

        try {
          // @ts-expect-error
          ({code, stdout, stderr} = await run(`npm`, `logout`, `--publish`, {
            env: {
              HOME: homePath,
              USERPROFILE: homePath,
              YARN_RC_FILENAME: SPEC_RC_FILENAME,
            },
          }));
        } catch (error) {
          ({code, stdout, stderr} = error);
        }

        await expect(yarn.readConfiguration(homePath, {
          filename: SPEC_RC_FILENAME,
        })).resolves.toStrictEqual({
          npmRegistries: {
            [FAKE_REGISTRY_URL]: FAKE_REGISTRY_CREDENTIALS,
            [FAKE_THIRD_REGISTRY_URL]: FAKE_REGISTRY_CREDENTIALS,
          },
        });

        expect({code, stdout, stderr}).toMatchSnapshot();
      })
    );

    it(
      `should logout a user from the scope registry when the -s,--scope flag is used`,
      makeTemporaryEnv({}, {
        npmRegistryServer: FAKE_REGISTRY_URL,
        npmPublishRegistry: FAKE_PUBLISH_REGISTRY_URL,
      }, async ({path, run, source}) => {
        const homePath = await xfs.mktempPromise();

        await yarn.writeConfiguration(homePath, {
          npmScopes: {
            [FAKE_FIRST_SCOPE]: FAKE_REGISTRY_CREDENTIALS,
            [FAKE_SECOND_SCOPE]: FAKE_REGISTRY_CREDENTIALS,
          },
        }, {
          filename: SPEC_RC_FILENAME,
        });

        let code: number;
        let stdout: string;
        let stderr: string;

        try {
          // @ts-expect-error
          ({code, stdout, stderr} = await run(`npm`, `logout`, `--scope`, FAKE_FIRST_SCOPE, {
            env: {
              HOME: homePath,
              USERPROFILE: homePath,
              YARN_RC_FILENAME: SPEC_RC_FILENAME,
            },
          }));
        } catch (error) {
          ({code, stdout, stderr} = error);
        }

        await expect(yarn.readConfiguration(homePath, {
          filename: SPEC_RC_FILENAME,
        })).resolves.toStrictEqual({
          npmScopes: {
            [FAKE_SECOND_SCOPE]: FAKE_REGISTRY_CREDENTIALS,
          },
        });

        expect({code, stdout, stderr}).toMatchSnapshot();
      })
    );
  });
});
