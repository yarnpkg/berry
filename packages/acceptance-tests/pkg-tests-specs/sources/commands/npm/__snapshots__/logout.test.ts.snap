// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Commands npm logout should logout a user from all registries and all scopes when the -A,--all flag is used 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: Successfully logged out from everything
➤ YN0000: Successfully logged out from http://yarn.test.registry
➤ YN0000: Done
",
}
`;

exports[`Commands npm logout should logout a user from the default registry when no arguments are passed 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: Successfully logged out from http://yarn.test.registry
➤ YN0000: Done
",
}
`;

exports[`Commands npm logout should logout a user from the publish registry when the --publish flag is used 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: Successfully logged out from https://npm.pkg.github.com
➤ YN0000: Done
",
}
`;

exports[`Commands npm logout should logout a user from the scope registry when the -s,--scope flag is used 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: Successfully logged out from first
➤ YN0000: Done
",
}
`;
